src/client/datascience/notebook/kernelProvider.ts

```diff
@@ -50,6 +50,8 @@ import { traceInfo, traceInfoIf } from '../../common/logger';
 
 @injectable()
 export class VSCodeKernelPickerProvider implements INotebookKernelProvider {
+    // Keep a mapping of Document Uri => NotebookCommunication for widget tests
+    public webviews = new Map<string, NotebookCommunication[]>();
```

Comment:
Basically just a move of this webview code from NotebookProvider to KernelProvider.

Reply:
Can we change this to use a WeakMap and index it with the `NotebookDocument`
I.e. `new WeakMap<NotebookDocument, NotebookCommunication>`
Feels better than using a string and also if the notebook is disposed, we don't need to remove the item from the map either.